bundles/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/SonyAuthChecker.java

```diff
@@ -0,0 +1,99 @@
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.sony.internal;
+
+import java.util.Objects;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.sony.internal.net.NetUtil;
+import org.openhab.binding.sony.internal.scalarweb.ScalarWebConstants;
+import org.openhab.binding.sony.internal.transports.SonyTransport;
+import org.openhab.binding.sony.internal.transports.TransportOptionHeader;
+
+/**
+ * This class contains the logic to determine if an authorization call is needed (via {@link SonyAuth})
+ * 
+ * @author Tim Roberts - Initial contribution
+ */
+@NonNullByDefault
+public class SonyAuthChecker {
+    /** The transport to use for check authorization */
+    private final SonyTransport transport;
+
+    /** The current access code */
+    private final @Nullable String accessCode;
+
+    /**
+     * Constructs the checker from the transport and access code
+     * 
+     * @param transport a non-null transport
+     * @param accessCode a possibly null, possibly empty access code
+     */
+    public SonyAuthChecker(final SonyTransport transport, final @Nullable String accessCode) {
+        Objects.requireNonNull(transport, "transport cannot be null");
+
+        this.transport = transport;
+        this.accessCode = accessCode;
+    }
+
+    /**
+     * Checks the result using the specified callback
+     * 
+     * @param callback a non-null callback
+     * @return a non-null result
+     */
+    public CheckResult checkResult(final CheckResultCallback callback) {
+        Objects.requireNonNull(callback, "callback cannot be null");
+
+        // Check to see if auth
+        final String localAccessCode = accessCode;
+        if (localAccessCode != null
+                && !StringUtils.equalsIgnoreCase(ScalarWebConstants.ACCESSCODE_RQST, localAccessCode)) {
+            final TransportOptionHeader authHeader = new TransportOptionHeader(
+                    NetUtil.createAccessCodeHeader(localAccessCode));
+            try {
+                transport.setOption(authHeader);
+                if (AccessResult.OK.equals(callback.checkResult())) {
+                    return CheckResult.OK_HEADER;
+                }
+            } finally {
+                transport.removeOption(authHeader);
+            }
+        }
+
+        final AccessResult res = callback.checkResult();
+        if (res == null) {
+            return new CheckResult(CheckResult.OTHER, "Check result returned null");
+        }
+
+        if (AccessResult.OK.equals(res)) {
+            return CheckResult.OK_COOKIE;
+        }
+        return new CheckResult(res);
+    }
```

Comment:
I don't see why you can't cache the result of your initial call to `callback.checkResult()`. But even if you cached it why does its execution need to be delayed? And if it doesn't need to be delayed why do you need a callback at all?
This whole thing doesn't make sense to me. Here are some ways to fix it:
1. Changing this into some abstract parent class/interface for the protocols and then use an abstract method as your callback.
2. Turn this into a utility class by making all the methods static, and get rid of the callback entirely.
3. Document the hell out this class. (My least favorite since I prefer code readability)

Reply:
Authentication is the hardest part of this because each line (and really some products) vary slightly.  What you are looking at is about the 100th rewrite of it and this is a part I will NOT touch because it's working now over everything and every single line is important and has been battle tested (catches some weird issue with some product somewhere).

Can't be cached because different products react differently (based on the product line AND how the user has setup authentication on the product itself) when you pass access code header and when you don't.  So worst case scenario is we try it with an access code, get rejected then try without the access code.

I'll try to document it a bit more although I'll confess - there are simply too many variables in play here to properly document it (depends on what we are trying to do [authenticate or not], what we were setup to do [access code, static access code or none], which product line is being used and how that product has been setup]).